File: Practica 3/p2.java
public class Cola {
    private long[] arreglo;  
    private int inicio;      
    private int fin;         
    private int n;           

    public Cola(int n) {
        this.n = n;
        this.arreglo = new long[n];
        this.inicio = 0;
        this.fin = -1;  
    }

    public void insert(long e) {
        if (isFull()) {
            throw new IllegalStateException("La cola está llena");
        }
        fin = (fin + 1) % n;  
        arreglo[fin] = e;
    }

    public long remove() {
        if (isEmpty()) {
            throw new IllegalStateException("La cola está vacía");
        }
        long elemento = arreglo[inicio];
        if (inicio == fin) {  
            inicio = 0;
            fin = -1;
        } else {
            inicio = (inicio + 1) % n;  
        }
        return elemento;
    }

    public long peek() {
        if (isEmpty()) {
            throw new IllegalStateException("La cola está vacía");
        }
        return arreglo[inicio];
    }

    public boolean isEmpty() {
        
        return fin == -1;
    }

    public boolean isFull() {
        
        
        return !isEmpty() && (fin + 1) % n == inicio;
    }

    public int size() {
        if (isEmpty()) {
            return 0;
        }
    
        if (fin >= inicio) {
            return fin - inicio + 1;
        } else {
            return n - inicio + fin + 1;
        }
    }

    public static void main(String[] args) {
        Cola cola = new Cola(5);
        cola.insert(10);
        cola.insert(20);
        cola.insert(30);
        System.out.println(cola.remove());  // Salida: 10
        System.out.println(cola.peek());    // Salida: 20
        System.out.println(cola.isEmpty()); // Salida: false
        System.out.println(cola.isFull());  // Salida: false
        System.out.println(cola.size());    // Salida: 2
    }
}

File: Practica 3/p2.py
class Cola:
    def __init__(self, n):
        self.arreglo = [None] * n  
        self.inicio = 0         
        self.fin = -1              
        self.n = n                 

    def insert(self, e):
        if self.isFull():
            raise Exception("La cola está llena")
        self.fin = (self.fin + 1) % self.n  
        self.arreglo[self.fin] = e

    def remove(self):
        if self.isEmpty():
            raise Exception("La cola está vacía")
        elemento = self.arreglo[self.inicio]
        if self.inicio == self.fin:  
            self.inicio = 0
            self.fin = -1
        else:
            self.inicio = (self.inicio + 1) % self.n  
        return elemento

    def peek(self):
        if self.isEmpty():
            raise Exception("La cola está vacía")
        return self.arreglo[self.inicio]

    def isEmpty(self):
        return self.fin == -1

    def isFull(self):
        
        return not self.isEmpty() and (self.fin + 1) % self.n == self.inicio

    def size(self):
        if self.isEmpty():
            return 0
        
        if self.fin >= self.inicio:
            return self.fin - self.inicio + 1
        else:
            return self.n - self.inicio + self.fin + 1
        
cola = Cola(5)
cola.insert(10)
cola.insert(20)
cola.insert(30)
print(cola.remove())  # Salida: 10
print(cola.peek())   # Salida: 20
print(cola.isEmpty()) # Salida: False
print(cola.isFull())  # Salida: False
print(cola.size())    # Salida: 2